Bring code up to cam6_4_100, resolve inconsistencies in ext libs, fix IC creation and restart issues

jtruesdal opened the pull request:
This PR fixes issues #51 and #58.  An inconsistent set of externals resulted in a compilation error for any compset using CTSM.  Additionally it fixes some critical bugs to allow restart testing and IC generation.  I updated the CAM externals to the set used in the latest alpha tag cesm3_0_alpha07c. Since that alpha tag used cam6_4_100 I've also updated the stormspeed CAM version. I've used a fork of the ccs_config_cesm1.0.57 to keep the external versions consistent with the alpha tag.  After this PR I'll look into making the non hydrostatic changes in a follow on PR to ccs_config.

The bug fixes and inconsistency mods are in .gitmodules, dyn_comp.F90, stepon.F90 and dp_coupling.F90.  The rest of the 70 some files were modified automatically when CAM was updated to cam6_4_100.  I ran a few bob start/restart and IC creation runs to make sure everything compiled and ran correctly.

closes Issue #51
closes Issue #58

peverwhee committed:
use ccppized init

peverwhee committed:
merge to 060

peverwhee committed:
Merge remote-tracking branch 'ESCOMP/cam_development' into rrtmgp-lw

peverwhee committed:
Merge remote-tracking branch 'ESCOMP/cam_development' into rrtmgp-lw

peverwhee committed:
rrtmgp_inputs incorporation validated

peverwhee committed:
lw cloud optics mostly done

peverwhee committed:
commit everything; will disentangle what goes into ccpp-physics later

peverwhee committed:
rrtmgp lw works!

peverwhee committed:
add initial attempt at type wrappers; code runs but answers are wrong

peverwhee committed:
finish object wrappers; answers now match again

peverwhee committed:
remove duplicate code; add in fix from cam_development

peverwhee committed:
merge to head of cam_development

Commit:
code clean-up and adding comments

peverwhee committed:
move ccppized schemes to atmospheric_physics; fix indexing issues; clean-up cesm-log logging; add rrtmgp paths to configure

peverwhee committed:
update atmos_phys hash

peverwhee committed:
update interface to calculate current calendar day on "host" side

peverwhee committed:
remove unused variable

peverwhee committed:
merge to head of cam_development

peverwhee committed:
subset arrays in call to ccpp layers

peverwhee committed:
handle conditionally present pbuf fields

peverwhee committed:
fix interface bug; remove unnecessary argument from cloud optics scheme

peverwhee committed:
update atmospheric physics hash

peverwhee committed:
update hash; add interstitial q variables

peverwhee committed:
merge to head of cam_development

peverwhee committed:
Merge remote-tracking branch 'ESCOMP/cam_development' into rrtmgp-lw

peverwhee committed:
update hash

fvitt committed:
An alternative for TEM diagnostics

        modified:   bld/namelist_files/namelist_definition.xml
        new file:   cime_config/testdefs/testmods_dirs/cam/outfrq9s_ctem/shell_commands
        new file:   cime_config/testdefs/testmods_dirs/cam/outfrq9s_ctem/user_nl_cam
        new file:   cime_config/testdefs/testmods_dirs/cam/outfrq9s_ctem/user_nl_clm
        modified:   src/control/runtime_opts.F90
        modified:   src/physics/cam/physpkg.F90
        modified:   src/physics/cam7/physpkg.F90
        new file:   src/utils/ctem_diags_mod.F90
        new file:   src/utils/esmf_check_error_mod.F90
        new file:   src/utils/esmf_lonlat_grid_mod.F90
        new file:   src/utils/esmf_phys2lonlat_mod.F90
        new file:   src/utils/esmf_phys_mesh_mod.F90
        new file:   src/utils/esmf_zonal_mean_mod.F90

Commit:
Merge remote-tracking branch 'ESCOMP/cam_development' into rrtmgp-lw

cacraigucar committed:
Update externals to match cesm3_0_beta06

cacraigucar committed:
update externals for git-fleximod

brian-eaton committed:
update default pes for mpasa120 on derecho

Commit:
address review comments

brian-eaton committed:
remove unused testmods outfrq3s_refined

fvitt committed:
Changes suggested by code reviewer

        modified:   src/physics/cam/physpkg.F90
        modified:   src/physics/cam7/physpkg.F90
        modified:   src/utils/ctem_diags_mod.F90
        modified:   src/utils/esmf_lonlat_grid_mod.F90
        modified:   src/utils/esmf_phys2lonlat_mod.F90
        modified:   src/utils/esmf_phys_mesh_mod.F90
        modified:   src/utils/esmf_zonal_mean_mod.F90

fvitt committed:
Fix typos

	modified:   src/utils/ctem_diags_mod.F90
	modified:   src/utils/esmf_lonlat_grid_mod.F90

brian-eaton committed:
test updates

brian-eaton committed:
update ChangeLog

cacraigucar committed:
Fix cosp test and final ChangeLog for cam6_4_096

cacraigucar committed:
Merge pull request #1321 from cacraigucar/cam_alpha06g

cam6_4_096: Update externals to match cesm3_0_beta06/alpha06g

fvitt committed:
Exercise the TEM diags in an aux_cam test

	modified:   cime_config/testdefs/testlist_cam.xml
	modified:   cime_config/testdefs/testmods_dirs/cam/outfrq9s_ctem/user_nl_cam

fvitt committed:
Merge remote-tracking branch 'ESCOMP/cam_development' into esmf_tem_diags

fvitt committed:
ChangeLog draft

fvitt committed:
ChangeLog update

fvitt committed:
Merge pull request #1319 from fvitt/esmf_tem_diags

cam6_4_097: An alternative for TEM diagnostics

peverwhee committed:
subset arguments to ncol

peverwhee committed:
address review comments

peverwhee committed:
update to latest atmos_phys hash; merge forward

peverwhee committed:
take another stab at the open acc directives

Commit:
fixes to appease the nag compiler; update atmos_phys hash

peverwhee committed:
update acc directives; remove redundant lowercasing

peverwhee committed:
Merge remote-tracking branch 'origin/rrtmgp-lw' into rrtmgp-lw

Commit:
address reviewer comments

Commit:
update git-fleximod

brian-eaton committed:
update to cam6_4_097

brian-eaton committed:
add FHIST_C4

brian-eaton committed:
refactor setting stream_ndep_year_* variables

brian-eaton committed:
testing updates

peverwhee committed:
update to atmospheric_physics tag

peverwhee committed:
final changelog

peverwhee committed:
Merge pull request #1290 from peverwhee/rrtmgp-lw

cam6_4_098: CAM updates to bring in CCPP-ized RRTMGP longwave modules

brian-eaton committed:
address review comments

brian-eaton committed:
update to cam6_4_098

brian-eaton committed:
fix issues 1326 and 1332; update ChangeLog

brian-eaton committed:
Merge pull request #1331 from brian-eaton/tests04

cam6_4_099: New F compset for cam4; updates to CAM testing.

peverwhee committed:
update rrtmgp interfaces

peverwhee committed:
update atmospheric_physics hash

peverwhee committed:
add changelog

peverwhee committed:
Merge pull request #1333 from peverwhee/fix-memory-leak

cam6_4_100: Memory leak fixes

jtruesdal committed:
Bring stormspeed up to 'cam6_4_100' and get externals in sync with Jian's cmeps changes.

Merge pull request #1333 from peverwhee/fix-memory-leak

jtruesdal committed:
update topo defaults for ne120 resolutions.

jtruesdal committed:
last minute bug fixes and some code cleanup

jtruesdal committed:
accidentally included some unchecked refactor code, removed

jtruesdal commented:
@sjsprecious Here is the update to stormspeed.  Lots of files but the only code I change is listed in the description above.  The rest of the changes are due to updating to cam6_4_100 to be consistent with the alpha tag set of externals.  After this is merged I'll update the refmesh mods and ask for your re-review on that PR.

sjsprecious commented on src/dynamics/senh/stepon.F90:
```diff
@@ -783,81 +748,35 @@ subroutine diag_dynvar_ic(elem)
+         end do
       end do
 
 !!$      if (fv_nphys > 0) then
```
remove this line if not needed.

sjsprecious commented on src/dynamics/senh/stepon.F90:
```diff
@@ -783,81 +748,35 @@ subroutine diag_dynvar_ic(elem)
-!!$
-!!$         deallocate(fld_fvm)
-!!$         deallocate(fld_gll)
 !!$      end if
```
remove this line if not needed.

sjsprecious commented on src/dynamics/senh/dyn_comp.F90:
```diff
@@ -756,53 +756,86 @@ end subroutine dyn_register
+    end if
 
-   call set_phis(dyn_in)
+    !jt   ! Create mapping files using SE basis functions if requested
+    !jt   call create_native_mapping_files(par, elem, 'native')
+    !jt   call create_native_mapping_files(par, elem, 'bilin')
```
remove these lines if not needed.

sjsprecious reviewed (changes requested):
Thanks @jtruesdal for working on these bug fixes and CAM tag update.

I only have some minor comments but it looks like the CI tests failed with a compilation error. Can you take a look and resolve them?

jtruesdal committed:
Fix dyn_comp again and some PR cleanup.

jtruesdal commented:
Updated dyn_comp.F90 so it is building and cleaned up some commented code. I'll have to take some time now and go back through the remaining comments and start checking if that code needs to be incorporated.  SCM functionality is some of the code commented out so will work on that soon.  

 **_I noticed some of the CI tests failing because it was looking for datasets in /glade/inputdata instead of /glade/p/cesmdata/inputdata._**. Is this an error in the code or the CI scripts?  I don't think I changed the file that a few of the tests are missing but I guess it could have been updated with the move to cam6_4_100.

sjsprecious commented:
> Updated dyn_comp.F90 so it is building and cleaned up some commented code. I'll have to take some time now and go back through the remaining comments and start checking if that code needs to be incorporated. SCM functionality is some of the code commented out so will work on that soon.

Thanks @jtruesdal for fixing this quickly.
 
> **_I noticed some of the CI tests failing because it was looking for datasets in /glade/inputdata instead of /glade/p/cesmdata/inputdata._**. Is this an error in the code or the CI scripts? I don't think I changed the file that a few of the tests are missing but I guess it could have been updated with the move to cam6_4_100.

The `/glade/p/cesmdata` path on Derecho is mounted to the `/glade` path in the container, so this should be fine. I think the problem is that the cloud runner can not find those files somehow and therefore it tried to download them. This is not possible since the cloud runner only has read access to the GLADE filesystem. I checked the `/glade/p/cesmdata` path on Derecho and files are there. But the date is quite recent so I think you are right that they may be updated and we need to refresh the path mounted to the cloud runner (not sure why we need to). I tested your PR manually on Derecho and it ran fine. Thus I will approve it and let you decide whether to merge it quickly or wait until the admin fixes the path/file issue (I will talk to him on Monday).

sjsprecious reviewed (approved):
Thanks @jtruesdal for addressing my comments. I will talk with the cloud admin about the path/file issue for the input data.

I see this git-fleximod test fails as well (https://github.com/NCAR/StormSPEED/actions/runs/17692143350/job/50287146200?pr=59) and I do not know if you can fix it or not. But it does not seem to touch any code changes here.
